src/theory/bags/bags_rewriter.h

```diff
@@ -0,0 +1,193 @@
+/*********************                                                        */
+/*! \file bags_rewriter.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Mudathir Mohamed
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Bags theory rewriter.
+ **/
+
+#include "cvc4_private.h"
+#include "theory/bags/rewrites.h"
+
+#ifndef CVC4__THEORY__BAGS__THEORY_BAGS_REWRITER_H
+#define CVC4__THEORY__BAGS__THEORY_BAGS_REWRITER_H
+
+#include "theory/rewriter.h"
+
+namespace CVC4 {
+namespace theory {
+namespace bags {
+
+struct BagsRewriteResponse
+{
+  RewriteStatus d_status;
+  Node d_node;
+  Rewrite d_rewrite;
+  BagsRewriteResponse();
+  BagsRewriteResponse(RewriteStatus status, Node n, Rewrite rewrite);
+  BagsRewriteResponse(const BagsRewriteResponse& r);
+  RewriteResponse getResponse() const;
+}; /* struct BagsRewriteResponse */
+
+class BagsRewriter : public TheoryRewriter
+{
+ public:
+  BagsRewriter(HistogramStat<Rewrite>* statistics = nullptr);
+
+  /**
+   * postRewrite nodes with kinds: MK_BAG, BAG_COUNT, UNION_MAX, UNION_DISJOINT,
+   * INTERSECTION_MIN, DIFFERENCE_SUBTRACT, DIFFERENCE_REMOVE, BAG_CHOOSE,
+   * BAG_CARD, BAG_IS_SINGLETON.
+   * See the rewrite rules for these kinds below.
+   */
+  RewriteResponse postRewrite(TNode n) override;
+  /**
+   * preRewrite nodes with kinds: EQUAL, BAG_IS_INCLUDED.
+   * See the rewrite rules for these kinds below.
+   */
+  RewriteResponse preRewrite(TNode n) override;
+
+ private:
+  /**
+   * patterns for n
+   * - (= A A) = true where A is a bag
+   */
+  BagsRewriteResponse rewriteEqual(const TNode& n) const;
+
+  /**
+   * patterns for n
+   * - (bag.is_included A B) = ((difference_subtract A B) == emptybag)
+   */
+  BagsRewriteResponse rewriteIsIncluded(const TNode& n) const;
+
+  /**
+   * patterns for n:
+   * - (mkBag x 0) = (emptybag T) where T is the type of x
```

Comment:
I would maybe keep this documentation elsewhere.

The problems is that rewrites may evolve rapidly and it is hard to keep documentation like this up to date.

Reply:
I would argue the mental effort to understand the rule without documentation later surpasses the painful effort of updating the documentation :)
Also writing the documentation makes it harder to make errors like `n[0][0]` when you meant `n[0][1]`.

Reply:
Ok, I would make the documentation less specific e.g. "patterns for rewrites include...".